flocker/node/agents/ebs.py

```diff
@@ -42,6 +47,141 @@
+    sets_attach = field(mandatory=True, type=bool)
+    unsets_attach = field(mandatory=True, type=bool)
+
+
+class VolumeStateTable(PRecord):
+    """
+    Map of volume operation to expected volume state transitions
+    and expected update to volume's ``attach_data``.
+    """
+
+    def _populate_volume_state_table():
+        """
+        Initialize volume state table  with transitions for ``create_volume``,
+        ``attach_volume``, ``detach_volume``, ``delete_volume`` operations.
+        """
+        O = VolumeOperations
+        S = VolumeStates
+        table = pmap()
+
+        def add_flow(operation, start, transient, end, sets_attach,
+                     unsets_attach):
+            """
+            Helper to add expected volume states for given operation.
+            """
+            return table.set(operation,
+                             VolumeStateFlow(start_state=start,
+                                             transient_state=transient,
+                                             end_state=end,
+                                             sets_attach=sets_attach,
+                                             unsets_attach=unsets_attach))
+
+        table = add_flow(O.CREATE, S.EMPTY, S.CREATING, S.AVAILABLE,
+                         False, False)
+        table = add_flow(O.ATTACH, S.AVAILABLE, S.ATTACHING, S.IN_USE,
+                         True, False)
+        table = add_flow(O.DETACH, S.IN_USE, S.DETACHING, S.AVAILABLE,
+                         False, True)
+        table = add_flow(O.DESTROY, S.AVAILABLE, S.DELETING, S.EMPTY,
+                         False, False)
+        return table
+
+    table = pmap_field(NamedConstant, VolumeStateFlow,
+                       initial=_populate_volume_state_table())
+
+VOLUME_STATE_TABLE = VolumeStateTable()
+
+
+class TimeoutException(Exception):
+    """
+    A timeout on waiting for volume to reach destination end state.
+
+    :param unicode blockdevice_id: Unique identifier for a volume.
+    :param NamedConstant operation: Operation performed on volume.
+    :param unicode start_state: Volume's start state before operation.
+    :param unicode transient_state: Expected transient state during operation.
+    :param unicode end_state: Expected end state on operation completion.
+    :param unicode current_state: Volume's state at timeout.
+    """
+    def __init__(self, blockdevice_id, operation,
+                 start_state, transient_state, end_state, current_state):
+        Exception.__init__(self, blockdevice_id)
+        self.blockdevice_id = blockdevice_id
+        Exception.__init__(self, operation)
+        self.operation = operation
+        Exception.__init__(self, start_state)
+        self.start_state = start_state
+        Exception.__init__(self, transient_state)
+        self.transient_state = transient_state
+        Exception.__init__(self, end_state)
+        self.end_state = end_state
+        Exception.__init__(self, current_state)
```

Comment:
Thanks for pointing out! Leaving ``blockdevice_id`` as the message, and rest of the parameters as state. Thanks!